Always check CRYPTO_LOCK_{read,write}_lock

richsalz opened the pull request:
Some functions that lock things are void, so we just return early.

Fixes: #14230

Also marks the lock routines `__owur` so the error doesn't propagate.

slontis commented on crypto/engine/eng_table.c:
```diff
@@ -161,7 +163,9 @@ IMPLEMENT_LHASH_DOALL_ARG(ENGINE_PILE, ENGINE);
 void engine_table_unregister(ENGINE_TABLE **table, ENGINE *e)
 {
-    CRYPTO_THREAD_write_lock(global_engine_lock);
+    if (!CRYPTO_THREAD_write_lock(global_engine_lock))
```
This is very unfortunate.. Same with the next one.

slontis replied:
Maybe assert here also?

richsalz replied:
there are several places where locks are done inside void functions.  I stopped  adding comments.  There's an open issue #14239.  Should I put the  comment everywhere so they will stick out?  A TODO(void) comment?

As for the assert, it could be here or perhaps pushed down to the actual locking code as discussed below.

kroeckx replied:
A TODO is fine for me.

richsalz replied:
created #14321.

slontis commented on crypto/objects/o_names.c:
```diff
@@ -207,17 +209,18 @@ int OBJ_NAME_add(const char *name, int type, const char *data)
-        goto unlock;
-    }
+    if (onp == NULL)
+        return 0;
```
This change doesn't look right?

richsalz replied:
No, the change is right because it would unlock something it didn't have.  The lock is taken a few lines down.

slontis commented on providers/fips/self_test.c:
```diff
@@ -216,9 +217,10 @@ static int verify_integrity(OSSL_CORE_BIO *bio, OSSL_FUNC_BIO_read_ex_fn read_ex
-    CRYPTO_THREAD_write_lock(fips_state_lock);
-    FIPS_state = state;
-    CRYPTO_THREAD_unlock(fips_state_lock);
+    if (ossl_assert(CRYPTO_THREAD_write_lock(fips_state_lock) != 0)) {
```
Doesn't CRYPTO_THREAD_write_lock() use ossl_assert already?

mattcaswell replied:
No. At least not in the pthreads implementation.

mattcaswell replied:
Well..at least there are some codepaths that call assert on failure but not others.

richsalz replied:
I think it makes sense to put ossl_assert into the locking calls.  What do you think?

mattcaswell replied:
The question is whether they could ever fail for reasons other than a programming error. If they can then assert is probably not appropriate. If not then it probably is. If the only reason that a locking call will ever fail is because of a programming error then I see no reason not to put the asserts in the locking calls themselves.

richsalz replied:
locking can fail because of lack of system resources. It is unlikely, but possible, just like malloc failing on other than embedded systems.

kroeckx replied:
Should it raise an error?

richsalz replied:
> should it raise an error?

I am going to say "not in this PR"  There is already #14321 to capture some future work thoughts.

kroeckx commented on crypto/context.c:
```diff
@@ -232,7 +232,12 @@ static void ossl_lib_ctx_generic_new(void *parent_ign, void *ptr_ign,
+            /*
+             * Can't return something, so best to hope that something will
+             * fail later. :(
+             */
```
Is there a reason you can't make this function return something?

richsalz replied:
it's a `new_func` function used in the crypto "exdata" stuff.  Look how "new_func" is used in `crypto/ex_data.c`  Changing it to non-void would be lots of work and subtly breaking things.

kroeckx commented on crypto/core_namemap.c:
```diff
@@ -107,7 +107,8 @@ int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
 
-    CRYPTO_THREAD_read_lock(namemap->lock);
+    if (!CRYPTO_THREAD_read_lock(namemap->lock))
+        return 0;
```
I'm not sure 0 is a good way to indicate an error.

richsalz replied:
> I'm not sure 0 is a good way to indicate an error.

but consistent with how the function is written.

kroeckx replied:
I'm not sure it is, I think the usage in ossl_namemap_stored() could use an error case.

richsalz replied:
Tell me what value you would like to see, @kroeckx,  and I'll make the change.  And then you can fight with @paulidale :)

paulidale replied:
@levitte wrote this, but I suggest: _e_<sup>&#120102;&#960;</sup>

kroeckx replied:
Works for me.

richsalz replied:
okay -1 it is.

kroeckx commented on crypto/core_namemap.c:
```diff
@@ -137,7 +138,8 @@ void ossl_namemap_doall_names(const OSSL_NAMEMAP *namemap, int number,
     cbdata.data = data;
-    CRYPTO_THREAD_read_lock(namemap->lock);
+    if (!CRYPTO_THREAD_read_lock(namemap->lock))
+        return;
```
This should probably be changed to indicate success or failure.

richsalz replied:
> This should probably be changed to indicate success or failure.

the "do_all" stuff doesn't support that sadly.

mattcaswell replied:
As of d84f5515f it does.

richsalz replied:
thanks for the update, @mattcaswell .  So should this function be changed to return an error?  If so, 0/1 or -1/1 ?

mattcaswell replied:
In latest master this function can fail (for other reasons besides locking) and either returns 0 for failure or 1 for success.

richsalz replied:
> In latest master this function can fail (for other reasons besides locking) and either returns 0 for failure or 1 for success.

turns out the code in this PR already does that, from when I recently rebased.

kroeckx commented on crypto/core_namemap.c:
```diff
@@ -170,7 +172,8 @@ int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
 
-    CRYPTO_THREAD_read_lock(namemap->lock);
+    if (!CRYPTO_THREAD_read_lock(namemap->lock))
+        return 0;
```
0 does not seem like a good way to return an error.

richsalz replied:
> 0 does not seem like a good way to return an error.

but it is how the function is written.  not my doing :)

kroeckx commented on crypto/core_namemap.c:
```diff
@@ -251,7 +254,8 @@ int ossl_namemap_add_name_n(OSSL_NAMEMAP *namemap, int number,
 
-    CRYPTO_THREAD_write_lock(namemap->lock);
+    if (!CRYPTO_THREAD_write_lock(namemap->lock))
+        return 0;
```
probably don't want 0

richsalz replied:
same reasoning.

kroeckx commented on crypto/engine/eng_pkey.c:
```diff
@@ -62,7 +62,8 @@ EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
     }
-    CRYPTO_THREAD_write_lock(global_engine_lock);
+    if (!CRYPTO_THREAD_write_lock(global_engine_lock))
+        return 0;
```
NULL

richsalz replied:
they were all '0' in the function so i changed them ALL to NULL.

kroeckx commented on crypto/engine/eng_pkey.c:
```diff
@@ -90,7 +91,8 @@ EVP_PKEY *ENGINE_load_public_key(ENGINE *e, const char *key_id,
     }
-    CRYPTO_THREAD_write_lock(global_engine_lock);
+    if (!CRYPTO_THREAD_write_lock(global_engine_lock))
+        return 0;
```
NULL

richsalz replied:
changed all 'return 0' in this function to 'return NULL'

kroeckx commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -156,7 +157,8 @@ void *evp_keymgmt_util_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
     }
 
-    CRYPTO_THREAD_write_lock(pk->lock);
+    if (!CRYPTO_THREAD_write_lock(pk->lock))
```
Should evp_keymgmt_freedata() get called?

richsalz replied:
> Should evp_keymgmt_freedata() get called?

good catch.  updated fixup commit pushed.

kroeckx commented on crypto/ex_data.c:
```diff
@@ -367,7 +368,8 @@ void CRYPTO_free_ex_data(int class_index, void *obj, CRYPTO_EX_DATA *ad)
         else {
-            CRYPTO_THREAD_write_lock(global->ex_data_lock);
+            if (!CRYPTO_THREAD_write_lock(global->ex_data_lock))
+                continue;
```
I'm unsure what the best way to handle this error is. We could also just ignore it. Maybe add some comment?

kroeckx replied:
Just leave it as it is.

richsalz commented:
rebased to fix conflicts.  will wait until you are done with the review @kroeckx .  Thanks for the incredibly detailed comments!

kroeckx commented on crypto/provider_core.c:
```diff
@@ -672,7 +674,8 @@ int ossl_provider_activate(OSSL_PROVIDER *prov)
     if (provider_activate(prov)) {
-        CRYPTO_THREAD_write_lock(prov->store->lock);
+        if (!CRYPTO_THREAD_write_lock(prov->store->lock))
+            return 0;
```
Should we call provider_deactivate()?

richsalz replied:
> Should we call provider_deactivate()?

yes! thanks.

kroeckx commented on crypto/rand/rand_lib.c:
```diff
@@ -204,7 +206,8 @@ int RAND_set_rand_engine(ENGINE *engine)
         }
     }
-    CRYPTO_THREAD_write_lock(rand_engine_lock);
+    if (!CRYPTO_THREAD_write_lock(rand_engine_lock))
```
Do we need to call ENGINE_finish() in case engine != NULL?

richsalz replied:
`ENGINE_finish` checks for NULL, but yes.

kroeckx commented on crypto/x509/by_dir.c:
```diff
@@ -268,7 +268,8 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
             htmp.hash = h;
-            CRYPTO_THREAD_read_lock(ctx->lock);
+            if (!CRYPTO_THREAD_read_lock(ctx->lock))
+                return 0;
```
goto finish?

richsalz replied:
goto finish, right!

kroeckx commented on crypto/x509/by_dir.c:
```diff
@@ -346,7 +347,8 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
         if (type == X509_LU_CRL) {
-            CRYPTO_THREAD_write_lock(ctx->lock);
+            if (!CRYPTO_THREAD_write_lock(ctx->lock))
+                return 0;
```
goto finish?

richsalz replied:
yes, goto finish.

kroeckx commented on ssl/ssl_sess.c:
```diff
@@ -319,8 +320,12 @@ int ssl_generate_session_id(SSL *s, SSL_SESSION *ss)
+        return 0;
+    if (!CRYPTO_THREAD_read_lock(s->session_ctx->lock)) {
+        CRYPTO_THREAD_unlock(s->lock);
+        return 0;
```
It should probably call SSLfatal()

richsalz replied:
added SSLfatal call, with `SSL_R_SESSION_ID_CONTEXT_UNINITIALIZED`  not perfect but i didn't want to add an error code.

richsalz commented:
Pushed a fixup commit that address all of @kroeckx 's wonderful detailed feedback.  (Not gonna lie: it was a pain going through and responding, but he was pretty much always right so it made this PR better.)

paulidale commented on crypto/engine/eng_dyn.c:
```diff
@@ -184,10 +184,12 @@ static int dynamic_set_data_ctx(ENGINE *e, dynamic_data_ctx **ctx)
      * context of the thread that won.
      */
+end:
     if (c)
```
Even though it's preexisting, could the `!= NULL` be added please.

richsalz replied:
> Even though it's preexisting, could the `!= NULL` be added please.

done.

paulidale reviewed (approved):
Assuming the required rebase goes cleanly, approved subject to one minor annoyance.

richsalz commented:
rebased to fix conflict ("Cert" vs "cert" in a comment so important :) and had to add a couple of additional cases to core_provider.  updated fixup commit pushed.

paulidale commented on crypto/provider_core.c:
```diff
@@ -203,8 +203,8 @@ int ossl_provider_disable_fallback_loading(OSSL_LIB_CTX *libctx)
-    if ((store = get_provider_store(libctx)) != NULL) {
-        CRYPTO_THREAD_write_lock(store->lock);
+    if ((store = get_provider_store(libctx)) != NULL
+            || !CRYPTO_THREAD_write_lock(store->lock)) {
```
This looks wrong.  It will modify the store without having the lock if the first condition is true.

richsalz replied:
"Rich, have you met my friend DeMorgan?"  Sigh.  Fixed.

paulidale reviewed (changes requested):
Oopsed it I think :(

paulidale commented on crypto/provider_core.c:
```diff
@@ -203,8 +203,8 @@ int ossl_provider_disable_fallback_loading(OSSL_LIB_CTX *libctx)
-    if ((store = get_provider_store(libctx)) != NULL) {
-        CRYPTO_THREAD_write_lock(store->lock);
+    if ((store = get_provider_store(libctx)) != NULL
+            && CRYPTO_THREAD_write_lock(store->lock)) {
```
I think this should be:

``` C
if ((store = get_provider_store(libctx)) != NULL) {
    if (!CRYPTO_THREAD_write_lock(store->lock))
        return 0;
    ...
}
```

The call has failed to do what it was meant to, so it ought to return an error.

richsalz replied:
the two fragments are equivalent, but your suggestion is more clear so I pushed an updated fixup.

paulidale commented:
Ping for review.

slontis commented on crypto/core_namemap.c:
```diff
@@ -107,7 +107,8 @@ int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
 
-    CRYPTO_THREAD_read_lock(namemap->lock);
+    if (!CRYPTO_THREAD_read_lock(namemap->lock))
+        return -1;
```
Why are you returning -1 here? Why not just 0.
You might want to search for usages of ossl_namemap_empty.
i.e if (namemap != NULL && ossl_namemap_empty(namemap)) {

richsalz replied:
I don't recall my reasoning, but it seems like 0/1 is only when you know the answer. Happy to change it if @paulidale (or other reviewer) doesn't mind.

paulidale replied:
I'm torn both ways -- both have advantages and disadvantages.

Outside of tests, there is only one use of this function and it is checking for empty.  This and the function's Boolean sounding name lead me to think that 0 on error might be the better choice.  The documentation in internal/man3 will need to be updated to note this.

kroeckx replied:
You can also make ossl_namemap_stored() return NULL in case of error.

richsalz replied:
additional commit pushed that returns 0 on error, fixed one place to handle that, and updated the docs in 82578da

paulidale commented on crypto/core_namemap.c:
```diff
@@ -108,7 +108,7 @@ int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
 
     if (!CRYPTO_THREAD_read_lock(namemap->lock))
-        return -1;
+        return 0;
```
Isn't this contrary to what you documented and it's use in `ossl_namemap_stored()`?

richsalz replied:
Arg.  Missed one.  Updated commit pushed.

slontis commented:
rebase time also..

richsalz commented:
rebased to fix small conflict

slontis commented:
looks like rebase pulled in a new change :) Error..

slontis commented on a.c:
```diff
@@ -0,0 +1,35 @@
+#define OPENSSL_SUPPRESS_DEPRECATED
```
I dont think you meant to include these 3 a.* files?

richsalz replied:
sigh.  commit pushed to remove them.

richsalz commented:
YARS  (Yet another rebase and squash) because someone added a locking call and the CI broke.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

paulidale commented:
CI isn't relevant.  Merged to master, thanks for the contribution.
